consensus/src/messages.rs

```
@@ -324,3 +332,121 @@ impl fmt::Debug for TC {
+        hasher.update(self.seed.digest());
+        Digest(hasher.finalize().as_slice()[..32].try_into().unwrap())
+    }
+}
+
+impl fmt::Debug for EpochChange {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(f, "EC({}, {:?})", self.since_epoch, self.seed)
+    }
+}
+
+/// Custom deserialize implementation to support compatibility with the older version of `Block` type.
+/// `epoch` field is not required in binary representation. If omitted, it is considered as Option::None
+impl<'de> Deserialize<'de> for Block {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        deserializer.deserialize_tuple(7, BlockVistitor)
+    }
+}
+
+struct BlockVistitor;
+
+impl<'de> serde::de::Visitor<'de> for BlockVistitor {
+    type Value = Block;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("expecting a Block object")
+    }
+
+    fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
+    where
+        V: serde::de::SeqAccess<'de>,
+    {
+        use serde::de::Error;
+        let qc: QC = seq
+            .next_element()?
+            .ok_or_else(|| Error::invalid_length(0, &self))?;
+        let tc: Option<TC> = seq
+            .next_element::<Option<TC>>()?
+            .ok_or_else(|| Error::invalid_length(1, &self))?;
+        let author = seq
+            .next_element()?
+            .ok_or_else(|| Error::invalid_length(2, &self))?;
+        let round = seq
+            .next_element()?
+            .ok_or_else(|| Error::invalid_length(3, &self))?;
+        let payload = seq
+            .next_element()?
+            .ok_or_else(|| Error::invalid_length(4, &self))?;
+        let signature = seq
+            .next_element()?
+            .ok_or_else(|| Error::invalid_length(5, &self))?;
+        let epoch = seq.next_element::<Option<Epoch>>().ok().flatten().flatten();
```

Interesting solution!
Didn't expect someone to think about this, or even implement it.

Just a question - in case the `Option<Epoch>` is malformed, this will fail silently and return `None`?

---

In this version - yes. 

We can improve it by allowing only `EOF` errors and treating them as `Option::None`, all others should be errors.

---

Let me know if you want me to add this change

---

No need, you answer is enough, thanks 🙂